### PR TITLE
Add scrolling options to PDFJS context menus.

### PR DESCRIPTION
**Pull request by sammygill**
#### d50f30c80ffad3e9a1fe97e3eb367a696a622683
<pre>
Add scrolling options to PDFJS context menus.
<a href="https://bugs.webkit.org/show_bug.cgi?id=237059">https://bugs.webkit.org/show_bug.cgi?id=237059</a>
rdar://89315125

Reviewed by Aditya Keerthi.

This patch adds scrolling options to the context menu when right
clicking on a PDF that is being rendered in PDFJS. It is important to
note that this patch only adds the options to the context menu in
appearance only. Linking the options to the actual implementation in
PDFJS will be done in another patch.

By taking advantage of the current context menu code, such as in
ContextMenuController, we can try to keep the PDF code as close to the
rest of the context menu code. For example, we would have have to go
through WebPageProxy::showPDFContextMenu inside WebPageProxyMac, which
the current PDFPlugin code does.

The only special consideration that needs to be made is within
ContextMenuController::populate. We need to be able to determine that
the PDF context menu items must be added to the context menu. We can
do this by determining that the current frame is an iframe
(!frame-&gt;isMainFrame) and that the owning document is a PDFDocument
(frame-&gt;ownerElement()-&gt;document().isPDFDocument()). Similarly, in
already existing cases where the code checks for frame-&gt;page(), we must
add an extra condition that the owning document is NOT a PDFDocument so
that we do not add any extra context menu items to PDF context menus and
remain consistent with the current PDF context menu.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
(WebCore::ContextMenuController::populate):
Goes through and determines where to add the PDF context menu items. We
need to make sure that the normal context menu remain the same (when
not viewing a PDF) and also to make sure that the PDF context menu
remains consistent with the current implementation.

* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
Adds the new context menu scrolling items to the ContextMenuAction
enum.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFSinglePage):
(WebCore::contextMenuItemPDFSinglePageContinuous):
(WebCore::contextMenuItemPDFTwoPages):
(WebCore::contextMenuItemPDFTwoPagesContinuous):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(toAction):
(toTag):

Canonical link: <a href="https://commits.webkit.org/253130@main">https://commits.webkit.org/253130@main</a>
</pre>

### CODE REVIEW

**Review by nt1m — CHANGES_REQUESTED** [review]
If this is specific to PDF.js, it should probably not touch PDFPlugin code

**Comment by sammygill** [comment]
> If this is specific to PDF.js, it should probably not touch PDFPlugin code

It looks like to get outside of PDFPlugin we will want to register some event handler outside of it. WebCore::callDefaultEventHandlersInBubblingOrder eventually calls HTMLPluginElement::defaultEventHandler(). Maybe this could be done inside WebCore::PDFDocument or is there perhaps a better area?

**Comment by sammygill** [comment]
> > If this is specific to PDF.js, it should probably not touch PDFPlugin code
> 
> It looks like to get outside of PDFPlugin we will want to register some event handler outside of it. WebCore::callDefaultEventHandlersInBubblingOrder eventually calls HTMLPluginElement::defaultEventHandler(). Maybe this could be done inside WebCore::PDFDocument or is there perhaps a better area?

Actually I don't think PDFDocument is the right place to do this. I'll need to look around a bit

**Comment by cdumez** [comment]
This PR title seems overly generic. It should probably say something about PDFjs or PDF.

**Comment by sammygill** [comment]
> This PR title seems overly generic. It should probably say something about PDFjs or PDF.

Updated!

File: Source/WebCore/page/ContextMenuController.cpp [hunk 1]
```diff
@@ -998,29 +1006,38 @@ void ContextMenuController::populate()
...
-                if (frame->page() && frame->page()->backForward().canGoBackOrForward(1))
-                    appendItem(ForwardItem, m_contextMenu.get());
-
-                // use isLoadingInAPISense rather than isLoading because Stop/Reload are
-                // intended to match WebKit's API, not WebCore's internal notion of loading status
-                if (loader.documentLoader()->isLoadingInAPISense())
-                    appendItem(StopItem, m_contextMenu.get());
-                else
-                    appendItem(ReloadItem, m_contextMenu.get());
+                if (frame->isMainFrame() || (frame->page() && !frame->ownerElement()->document().isPDFDocument())) {
```
Comment (cdumez) [inline Source/WebCore/page/ContextMenuController.cpp]: Did you mean to null check frame->ownerElement() instead of frame->page()?

File: Source/WebCore/page/ContextMenuController.cpp [hunk 2]
```diff
@@ -998,29 +1006,38 @@ void ContextMenuController::populate()
...
+                    if (loader.documentLoader()->isLoadingInAPISense())
+                        appendItem(StopItem, m_contextMenu.get());
+                    else
+                        appendItem(ReloadItem, m_contextMenu.get());
+                }
 #endif
                 }
 
-                if (frame->page() && !frame->isMainFrame())
+                if (frame->page() && !frame->isMainFrame() && !frame->ownerElement()->document().isPDFDocument())
```
Comment (cdumez) [inline Source/WebCore/page/ContextMenuController.cpp]: What guarantees that frame->ownerElement() is non null?

Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: I think I had some misunderstandings between frames and pages. I will definitely need to add a null check for that. If we got rid of the frame->page() check here would that be ok? Since a frame could get detached from a page (although I don't entirely know what that scenario would be), would we want to check that or do you not think it would matter in this case? Also, it looks like I might want to add a null check for the document as well?

Comment (cdumez) [reply Source/WebCore/page/ContextMenuController.cpp]: > might want to add a null check for the document as well?
No, document() is a reference, not a pointer so it cannot be null.

But when you dereference this pointer (frame->ownerElement()) without a null check, this looks dangerous.

File: Source/WebCore/page/ContextMenuController.cpp (hunk 1 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Added the null pointer check here. Also removed the frame->page() check since the more specific if branches are checking for them along with other checks below.

File: Source/WebCore/page/ContextMenuController.cpp (hunk 2 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Added the frame->ownerElement null pointer check here

File: Source/WebCore/platform/ContextMenuItem.h [hunk 3]
```diff
@@ -153,7 +153,10 @@ enum ContextMenuAction {
     ContextMenuItemLastNonCustomTag = ContextMenuItemTagCopySubject,
     ContextMenuItemBaseCustomTag = 5000,
     ContextMenuItemLastCustomTag = 5999,
-    ContextMenuItemBaseApplicationTag = 10000
+    ContextMenuItemBaseApplicationTag = 10000,
+    ContextMenuItemPDFSinglePageContinuous,
+    ContextMenuItemPDFTwoPages,
+    ContextMenuItemPDFTwoPagesContinuous
```
Comment (pxlcoder) [inline Source/WebCore/platform/ContextMenuItem.h]: These should be moved before `ContextMenuItemLastNonCustomTag`, and `ContextMenuItemLastNonCustomTag` should equal `ContextMenuItemPDFTwoPagesContinuous`.

**Comment by nt1m** [comment]
Can you guard the new code behind `ENABLE(PDFJS)` ?

File: Source/WebCore/page/ContextMenuController.cpp [hunk 4]
```diff
@@ -57,6 +57,7 @@
 #include "MouseEvent.h"
 #include "NavigationAction.h"
 #include "Node.h"
+#include "PDFDocument.h"
```
Comment (nt1m) [inline Source/WebCore/page/ContextMenuController.cpp]: You don't need to import this.

File: Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h [hunk 5]
```diff
@@ -117,6 +117,9 @@ enum {
     kWKContextMenuItemTagMediaPlayPause,
     kWKContextMenuItemTagMediaMute,
     kWKContextMenuItemTagDictationAlternative,
+    kWKContextMenuItemPDFSinglePageContinuous,
+    kWKContextMenuItemPDFTwoPages,
+    kWKContextMenuItemPDFTwoPagesContinuous,
```
Comment (whsieh) [inline Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h]: New API/SPI values should be appended to the end (in this case, after `kWKContextMenuItemTagCopyCroppedImage`), in order to preserve binary compatibility.

File: Source/WebKitLegacy/mac/WebView/WebUIDelegate.h [hunk 6]
```diff
@@ -73,6 +73,9 @@ enum {
     WebMenuItemPDFZoomOut,
     WebMenuItemPDFAutoSize,
     WebMenuItemPDFSinglePage,
+    WebMenuItemPDFSinglePageContinuous,
+    WebMenuItemPDFTwoPages,
+    WebMenuItemPDFTwoPagesContinuous,
```
Comment (whsieh) [inline Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: (Ditto, re: ordering of these new SPI enum values)

I'm also curious if we could just omit these as SPI-exposed values, altogether.

Comment (sammygill) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: I can do either of those. Is there a preference to which path we should take?

CC @pxlcoder

Comment (whsieh) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: > I can do either of those. Is there a preference to which path we should take?
> 
> CC @pxlcoder

I think it would be more ideal to avoid adding more legacy SPI, if possible.

Comment (pxlcoder) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: Agreed, note that this enum is marked as deprecated: `WEBKIT_ENUM_DEPRECATED_MAC`.

Comment (sammygill) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: `toTag` inside of `WebKitLegacy/Mac/WebView/WebHTMLView.mm` seems to be branching on `ContextMenuAction` values and is returning the appropriate enum from `WebUIDelegate` (which is deprecated as Aditya mentioned). Should we keep these values then or is there a specific value we can return for these cases?

Comment (pxlcoder) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: Can use the `return std::nullopt;` near the bottom.

(see `ContextMenuItemTagLookUpImage`)

File: Source/WebCore/page/ContextMenuController.cpp (hunk 4 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Removed

File: Source/WebCore/platform/ContextMenuItem.h (hunk 3 above)
Comment (sammygill) [reply Source/WebCore/platform/ContextMenuItem.h]: Moved up

File: Source/WebKitLegacy/mac/WebView/WebUIDelegate.h (hunk 6 above)
Comment (sammygill) [reply Source/WebKitLegacy/mac/WebView/WebUIDelegate.h]: I removed the mentioned enum values and also added a default case to return std:nullopt inside of `toTag`

File: Source/WebKitLegacy/mac/WebView/WebHTMLView.mm [hunk 7]
```diff
@@ -609,6 +609,9 @@ - (void)forwardContextMenuAction:(id)sender;
 
     case ContextMenuItemBaseApplicationTag:
         ASSERT_NOT_REACHED();
+    default:
+        return std::nullopt;
+        break;
```
Comment (whsieh) [inline Source/WebKitLegacy/mac/WebView/WebHTMLView.mm]: Nit - if we go with the `default:` case here, I think we can also remove some of the cases above where we currently return `std::nullopt`.

That said, I think there _might_ be an advantage to the (less succinct) approach of enumerating each of the new types, which is that it would force folks adding new item types in the future to consider (at compile-time) whether or not adding a corresponding `WebMenuItemTag*` is necessary. However, since this is legacy WebKit, the answer should probably be "no" anyways, so either way seems fine!

File: Source/WebCore/en.lproj/Localizable.strings [hunk 8]
```diff
@@ -1701,3 +1701,12 @@
 
 /* Lockdown Mode alert message */
 "Certain experiences and features may not function as expected. You can turn off Lockdown Mode for this app in Settings." = "Certain experiences and features may not function as expected. You can turn off Lockdown Mode for this app in Settings.";
+"Single Page" = "Single Page";
```
Comment (pxlcoder) [inline Source/WebCore/en.lproj/Localizable.strings:1706]: Nit: This should be moved under your `PDFJS strings` comment.

File: Source/WebCore/en.lproj/Localizable.strings [hunk 9]
```diff
@@ -1701,3 +1701,12 @@
...
+
+/** PDFJS strings*/
```
Comment (pxlcoder) [inline Source/WebCore/en.lproj/Localizable.strings]: Nit: Formatting: `/* PDFJS strings */`.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 10]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
-                    appendItem(ForwardItem, m_contextMenu.get());
-
-                // use isLoadingInAPISense rather than isLoading because Stop/Reload are
-                // intended to match WebKit's API, not WebCore's internal notion of loading status
-                if (loader.documentLoader()->isLoadingInAPISense())
-                    appendItem(StopItem, m_contextMenu.get());
-                else
-                    appendItem(ReloadItem, m_contextMenu.get());
+                if (frame->isMainFrame() || (frame->ownerElement() && !frame->ownerElement()->document().isPDFDocument())) {
+                    if (frame->page() && frame->page()->backForward().canGoBackOrForward(-1))
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: I know this is the existing code, but can we store `frame->page()` in a local variable while we're here?

File: Source/WebCore/page/ContextMenuController.cpp [hunk 11]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
+                        appendItem(BackItem, m_contextMenu.get());
+
+                    if (frame->page() && frame->page()->backForward().canGoBackOrForward(1))
+                        appendItem(ForwardItem, m_contextMenu.get());
+                    
+                    // use isLoadingInAPISense rather than isLoading because Stop/Reload are
+                    // intended to match WebKit's API, not WebCore's internal notion of loading status
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Comments should be full sentences with capitalization and punctuation.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 12]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
+                        appendItem(ReloadItem, m_contextMenu.get());
+                }
 #endif
                 }
 
-                if (frame->page() && !frame->isMainFrame())
+                if (frame->ownerElement() && !frame->isMainFrame() && !frame->ownerElement()->document().isPDFDocument()) 
                     appendItem(OpenFrameItem, m_contextMenu.get());
-
+
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Remove whitespace.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 13]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
                     appendItem(OpenFrameItem, m_contextMenu.get());
-
+                
                 if (!ShareMenuItem.isNull()) {
                     appendItem(*separatorItem(), m_contextMenu.get());
                     appendItem(ShareMenuItem, m_contextMenu.get());
                 }
             }
+#if ENABLE(PDFJS)
+            if (frame->ownerElement() && frame->ownerElement()->document().isPDFDocument()) {
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: This can be `if (auto* ownerElement = frame->ownerElement(); ownerElement->document().isPDFDocument())`.

File: Source/WebKitLegacy/mac/WebView/WebHTMLView.mm (hunk 7 above)
Comment (pxlcoder) [reply Source/WebKitLegacy/mac/WebView/WebHTMLView.mm]: I have a slight preference to explicitly add the new enum values. You could add them underneath `case ContextMenuItemTagLookUpImage:`.

Comment (sammygill) [reply Source/WebKitLegacy/mac/WebView/WebHTMLView.mm]: Removed the default case and added the new enum values below ContextMenuItemTagLookUpImage

File: Source/WebCore/en.lproj/Localizable.strings (hunk 8 above)
Comment (sammygill) [reply Source/WebCore/en.lproj/Localizable.strings:1706]: Moved

File: Source/WebCore/en.lproj/Localizable.strings (hunk 9 above)
Comment (sammygill) [reply Source/WebCore/en.lproj/Localizable.strings]: Fixed

File: Source/WebCore/page/ContextMenuController.cpp (hunk 10 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Added framePage variable

File: Source/WebCore/page/ContextMenuController.cpp (hunk 11 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Fixed

File: Source/WebCore/page/ContextMenuController.cpp (hunk 12 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Removed

File: Source/WebCore/page/ContextMenuController.cpp (hunk 13 above)
Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: Updated if statement

File: Source/WebCore/page/ContextMenuController.cpp [hunk 14]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
-                if (frame->page() && frame->page()->backForward().canGoBackOrForward(1))
-                    appendItem(ForwardItem, m_contextMenu.get());
-
-                // use isLoadingInAPISense rather than isLoading because Stop/Reload are
-                // intended to match WebKit's API, not WebCore's internal notion of loading status
-                if (loader.documentLoader()->isLoadingInAPISense())
-                    appendItem(StopItem, m_contextMenu.get());
-                else
-                    appendItem(ReloadItem, m_contextMenu.get());
+                if (frame->isMainFrame() || (frame->ownerElement() && !frame->ownerElement()->document().isPDFDocument())) {
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: I realized this check: `frame->ownerElement() && !frame->ownerElement()->document().isPDFDocument())` is performed 3 times in this method. Can we pull it out into a local variable, and reuse that?

File: Source/WebCore/platform/ContextMenuItem.h [hunk 15]
```diff
@@ -142,6 +142,11 @@ enum ContextMenuAction {
     ContextMenuItemTagMediaPlayPause,
     ContextMenuItemTagMediaMute,
     ContextMenuItemTagDictationAlternative,
+#if ENABLE(PDFJS)
+    ContextMenuItemPDFSinglePageContinuous,
+    ContextMenuItemPDFTwoPages,
+    ContextMenuItemPDFTwoPagesContinuous,
+#endif
```
Comment (pxlcoder) [inline Source/WebCore/platform/ContextMenuItem.h]: We need to add these values at the end for binary compatibility.

I also wonder if the `ENABLE(PDFJS)` is actually needed around these values. It might be clear to just omit the build flag. Any thoughts @whsieh?

Comment (nt1m) [reply Source/WebCore/platform/ContextMenuItem.h]: I think it would be better to keep the build flag, so it's easier to find/delete PDFJS related code.

Comment (sammygill) [reply Source/WebCore/platform/ContextMenuItem.h]: I guess the solution here would be to include ContextMenuItemLastNonCustomTag inside an if else macro? Inside the if it would point to the PDFJS enum and otherwise it would be as it is in the else branch?

Comment (pxlcoder) [reply Source/WebCore/platform/ContextMenuItem.h]: Yes, but using a build flag here would also cause binary compatibility issues if we added new values at the end, and then the value of the flag were to change.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 16]
```diff
@@ -998,29 +1006,39 @@ void ContextMenuController::populate()
...
+                    auto* framePage = frame->page();
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: This variable can be called `page`, we don't use the term `framePage` elsewhere in the code.

Comment (sammygill) [reply Source/WebCore/page/ContextMenuController.cpp]: I added 3 new local variables: ownerElement, isPDFDocument, and isMainFrame. For areas where we checked for both ownerElement and isPDFDocument, I changed it to just check for isPDFDocument since the value of that boolean also implicitly checks for ownerElement. If you think there might be something wrong with that logic let me know and I can add back in the ownerElement check.

File: Source/WebCore/platform/ContextMenuItem.h (hunk 15 above)
Comment (sammygill) [reply Source/WebCore/platform/ContextMenuItem.h]: I removed these build flags and moved the new enum values down. These values will now always be declared, but I still guarded sections of the code that used them.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 17]
```diff
@@ -965,6 +973,10 @@ void ContextMenuController::populate()
         if (selectionIsInsideImageOverlay || (linkURL.isEmpty() && mediaURL.isEmpty() && imageURL.isEmpty())) {
             if (!imageURL.isEmpty())
                 appendItem(*separatorItem(), m_contextMenu.get());
+            
+            auto* ownerElement = frame->ownerElement();
+            bool isPDFDocument = ownerElement ? ownerElement->document().isPDFDocument() : false;
```
Comment (cdumez) [inline Source/WebCore/page/ContextMenuController.cpp]: I think this would look nicer:
`bool isPDFDocument = ownerElement && ownerElement->document().isPDFDocument();`

File: Source/WebCore/page/ContextMenuController.cpp (hunk 16 above)
Comment (cdumez) [reply Source/WebCore/page/ContextMenuController.cpp]: I agree we should call this `page` rather than `framePage`

File: Source/WebCore/platform/ContextMenuItem.cpp [hunk 18]
```diff
@@ -197,6 +197,11 @@ bool isValidContextMenuAction(ContextMenuAction action)
     case ContextMenuAction::ContextMenuItemPDFZoomOut:
     case ContextMenuAction::ContextMenuItemPDFAutoSize:
     case ContextMenuAction::ContextMenuItemPDFSinglePage:
+#if ENABLE(PDFJS)
+    case ContextMenuAction::ContextMenuItemPDFSinglePageContinuous:
+    case ContextMenuAction::ContextMenuItemPDFTwoPages:
+    case ContextMenuAction::ContextMenuItemPDFTwoPagesContinuous:
+#endif
```
Comment (pxlcoder) [inline Source/WebCore/platform/ContextMenuItem.cpp]: I would remove the build flag around the cases, since they are defined everywhere, and instead, return `true` or `false` depending on its value:

```
case ContextMenuAction::ContextMenuItemPDFSinglePageContinuous:
case ContextMenuAction::ContextMenuItemPDFTwoPages:
case ContextMenuAction::ContextMenuItemPDFTwoPagesContinuous: {
#if ENABLE(PDFJS)
    return true;
#else
    return false;
#endif
}
```

File: Source/WebCore/page/ContextMenuController.cpp [hunk 19]
```diff
@@ -989,7 +1001,8 @@ void ContextMenuController::populate()
                 appendItem(SpeechMenuItem, m_contextMenu.get());
 #endif                
             } else {
-                if (!(frame->page() && (frame->page()->inspectorController().inspectionLevel() > 0 || frame->page()->inspectorController().hasRemoteFrontend()))) {
+                
+                if (!(frame->page() && (frame->page()->inspectorController().inspectionLevel() > 0 || frame->page()->inspectorController().hasRemoteFrontend()))) {
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Unnecessary whitespace.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 20]
```diff
@@ -998,29 +1011,41 @@ void ContextMenuController::populate()
...
+                    if (loader.documentLoader()->isLoadingInAPISense())
+                        appendItem(StopItem, m_contextMenu.get());
+                    else
+                        appendItem(ReloadItem, m_contextMenu.get());
+                }
 #endif
                 }
 
-                if (frame->page() && !frame->isMainFrame())
+                if (ownerElement && !isMainFrame && !isPDFDocument)
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Can you explain the reasoning behind this change? Should it be restricted to `ENABLE(PDFJS)`?

File: Source/WebCore/page/ContextMenuController.cpp [hunk 21]
```diff
@@ -989,7 +1002,7 @@ void ContextMenuController::populate()
                 appendItem(SpeechMenuItem, m_contextMenu.get());
 #endif                
             } else {
-                if (!(frame->page() && (frame->page()->inspectorController().inspectionLevel() > 0 || frame->page()->inspectorController().hasRemoteFrontend()))) {
+                if (!(page && (page->inspectorController().inspectionLevel() > 0 || page->inspectorController().hasRemoteFrontend()))) {
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Trailing whitespace.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 22]
```diff
@@ -965,6 +973,11 @@ void ContextMenuController::populate()
         if (selectionIsInsideImageOverlay || (linkURL.isEmpty() && mediaURL.isEmpty() && imageURL.isEmpty())) {
             if (!imageURL.isEmpty())
                 appendItem(*separatorItem(), m_contextMenu.get());
+            
+            auto* ownerElement = frame->ownerElement();
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Let's use `RefPtr` instead of `auto*`.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 23]
```diff
@@ -965,6 +973,11 @@ void ContextMenuController::populate()
...
+            bool isPDFDocument = ownerElement && ownerElement->document().isPDFDocument();
+            bool isMainFrame = frame->isMainFrame();
+            auto* page = frame->page();
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Let's use `RefPtr` instead of `auto*`, and move this variable up next to `ownerElement`.

File: Source/WebCore/platform/ContextMenuItem.cpp [hunk 24]
```diff
@@ -197,6 +197,7 @@ bool isValidContextMenuAction(ContextMenuAction action)
     case ContextMenuAction::ContextMenuItemPDFZoomOut:
     case ContextMenuAction::ContextMenuItemPDFAutoSize:
     case ContextMenuAction::ContextMenuItemPDFSinglePage:
+
```
Comment (pxlcoder) [inline Source/WebCore/platform/ContextMenuItem.cpp]: Nit: Please undo.

**Review by pxlcoder — APPROVED** [review]

File: Source/WebCore/page/ContextMenuController.cpp (hunk 20 above)
Comment (pxlcoder) [reply Source/WebCore/page/ContextMenuController.cpp]: I talked to @nt1m, and he informed me that `isPDFDocument` is only true for PDFJS documents, so this is fine.

File: Source/WebCore/page/ContextMenuController.cpp (hunk 23 above)
Comment (pxlcoder) [reply Source/WebCore/page/ContextMenuController.cpp]:
```
RefPtr page = frame->page();
RefPtr ownerElement = frame->ownerElement();
bool isPDFDocument = ownerElement && ownerElement->document().isPDFDocument();
bool isMainFrame = frame->isMainFrame();
```

File: Source/WebCore/page/ContextMenuController.cpp [hunk 25]
```diff
@@ -966,6 +974,12 @@ void ContextMenuController::populate()
             if (!imageURL.isEmpty())
                 appendItem(*separatorItem(), m_contextMenu.get());
 
+
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Unnecessary newline here.

File: Source/WebCore/page/ContextMenuController.cpp [hunk 26]
```diff
@@ -989,7 +1003,7 @@ void ContextMenuController::populate()
                 appendItem(SpeechMenuItem, m_contextMenu.get());
 #endif                
             } else {
-                if (!(frame->page() && (frame->page()->inspectorController().inspectionLevel() > 0 || frame->page()->inspectorController().hasRemoteFrontend()))) {
+                if (!(page && (page->inspectorController().inspectionLevel() > 0 || page->inspectorController().hasRemoteFrontend()))) {
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Trailing whitespace is still there.

File: Source/WebCore/en.lproj/Localizable.strings (hunk 8 above)
Comment (pxlcoder) [reply Source/WebCore/en.lproj/Localizable.strings:1706]: `/* PDFJS strings*/` -> `/* PDFJS strings */`

File: Source/WebCore/platform/ContextMenuItem.cpp [hunk 27]
```diff
@@ -265,6 +265,10 @@ bool isValidContextMenuAction(ContextMenuAction action)
     case ContextMenuAction::ContextMenuItemLastCustomTag:
     case ContextMenuAction::ContextMenuItemBaseApplicationTag:
         return true;
+    case ContextMenuAction::ContextMenuItemPDFSinglePageContinuous:
+    case ContextMenuAction::ContextMenuItemPDFTwoPages:
+    case ContextMenuAction::ContextMenuItemPDFTwoPagesContinuous:
+        return true;
```
Comment (pxlcoder) [inline Source/WebCore/platform/ContextMenuItem.cpp]: @sgill26 It would be good to move these up and share the `return true;` with the other cases in a follow-up patch (maybe the patch where you hook up the context menu items into PDF.js).

File: Source/WebCore/page/ContextMenuController.cpp [hunk 28]
```diff
@@ -811,6 +811,14 @@ void ContextMenuController::populate()
     ContextMenuItem SearchSpotlightItem(ActionType, ContextMenuItemTagSearchInSpotlight, 
         contextMenuItemTagSearchInSpotlight());
 #endif
+
+#if ENABLE(PDFJS)
+ContextMenuItem SinglePageItem(ActionType, ContextMenuItemPDFSinglePage, contextMenuItemPDFSinglePage());
+ContextMenuItem SinglePageContinuousItem(ActionType, ContextMenuItemPDFSinglePageContinuous, contextMenuItemPDFSinglePageContinuous());
+ContextMenuItem TwoPagesItem(ActionType, ContextMenuItemPDFTwoPages, contextMenuItemPDFTwoPages());
+ContextMenuItem TwoPagesContinuousItem(ActionType, ContextMenuItemPDFTwoPagesContinuous, contextMenuItemPDFTwoPagesContinuous());
```
Comment (pxlcoder) [inline Source/WebCore/page/ContextMenuController.cpp]: Nit: Indentation.
